File: python/grayt/spacetime.py
"""Spacetimes: the geometric layer of the ontology.

A ``Spacetime`` names a metric implemented in the Fortran core (module
SPACETIME) via an integer id ``mid`` and a parameter vector ``par``.
Analytic registered metrics use the Fortran dispatcher. CustomMetric and
TabulatedMetric (grayt.metrics) import compatible stationary axisymmetric
geometry without recompilation. Radiation and boundary choices are separate.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from . import _core

MID_KERR = 1
MID_QMETRIC = 2


class Spacetime:
    """Base class for spacetimes (geometrized units G = c = M = 1).

    Subclasses define ``mid`` (Fortran metric id) and ``par`` (length-4
    parameter vector) plus a ``capture_radius`` below which rays count
    as captured.
    """

    mid: int = 0

    @property
    def par(self) -> np.ndarray:
        raise NotImplementedError

    @property
    def capture_radius(self) -> float:
        raise NotImplementedError

    def metric_cov(self, r, theta):
        """Covariant components (tt, t-phi, rr, thth, phph) at (r, theta)."""
        return _core.spacetime.metric_cov(self.mid, self.par, r, theta)

    def metric_contra(self, r, theta):
        """Contravariant components and their r/theta derivatives."""
        return _core.spacetime.metric_contra(self.mid, self.par, r, theta)


@dataclass(frozen=True)
class BlackHole(Spacetime):
    """Kerr black hole with dimensionless spin ``a`` (Boyer-Lindquist)."""

    a: float = 0.0
    mid = MID_KERR

    def __post_init__(self):
        if not -1.0 <= self.a <= 1.0:
            raise ValueError(f"spin must satisfy |a| <= 1, got {self.a}")

    @property
    def par(self) -> np.ndarray:
        return np.array([self.a, 0.0, 0.0, 0.0])

    @property
    def horizon(self) -> float:
        """Outer event horizon radius r_H = 1 + sqrt(1 - a^2)."""
        return float(_core.raytracer.get_horizon(self.a))

    @property
    def isco(self) -> float:
        """Prograde innermost stable circular orbit radius."""
        return float(_core.raytracer.get_isco(self.a))

    @property
    def capture_radius(self) -> float:
        return self.horizon


@dataclass(frozen=True)
class QMetric(Spacetime):
    """q-metric (Zipoy-Voorhees): static, axisymmetric vacuum solution
    with quadrupole ``q`` (Appendix A of arXiv:2202.00086). q = 0 is
    Schwarzschild; for -1 < q < -1 + sqrt(3/2), q /= 0, the source is a
    naked singularity. The surface r = 2m is singular for q /= 0; rays
    are captured just outside it.
    """

    q: float = 0.0
    mid = MID_QMETRIC

    def __post_init__(self):
        if self.q <= -1.0:
            raise ValueError(f"quadrupole must satisfy q > -1, got {self.q}")

    @property
    def par(self) -> np.ndarray:
        return np.array([self.q, 0.0, 0.0, 0.0])

    @property
    def horizon(self) -> float:
        """The singular surface r = 2m (an event horizon only for q = 0);
        exposed under this name for plotting helpers."""
        return 2.0

    @property
    def capture_radius(self) -> float:
        return 2.0

File: src/spacetime.f90
MODULE SPACETIME
   ! Spacetime dispatcher: one entry point for metric evaluation, keyed
   ! by a metric id MID and a parameter vector PAR(4).
   !
   !   MID_KERR    = 1 : Kerr in Boyer-Lindquist, PAR(1) = spin a
   !   MID_QMETRIC = 2 : q-metric (Zipoy-Voorhees), PAR(1) = quadrupole q
   !
   ! Everything downstream (equations of motion, integrators, tracers)
   ! is metric-agnostic and reaches the geometry only through this
   ! module. Adding a spacetime = one new module + one CASE line here.
   USE KIND_PARAMS, ONLY: WP
   USE KERR_METRIC, ONLY: KERR_COV => METRIC_COV, &
                          KERR_CONTRA => METRIC_CONTRA, HORIZON_RADIUS
   USE Q_METRIC, ONLY: QMETRIC_COV, QMETRIC_CONTRA
   USE TABULATED_METRIC, ONLY: TABLE_COV, TABLE_CONTRA
   IMPLICIT NONE
   PRIVATE

   PUBLIC :: METRIC_COV, METRIC_CONTRA, INNER_BOUNDARY
   INTEGER, PARAMETER, PUBLIC :: MID_KERR = 1
   INTEGER, PARAMETER, PUBLIC :: MID_QMETRIC = 2

CONTAINS

   PURE SUBROUTINE METRIC_COV(MID, PAR, R, THETA, GD)
      INTEGER, INTENT(IN)   :: MID
      REAL(WP), INTENT(IN)  :: PAR(4), R, THETA
      REAL(WP), INTENT(OUT) :: GD(5)
      SELECT CASE (MID)
      CASE (3)
         CALL TABLE_COV(R, THETA, GD)
      CASE (MID_QMETRIC)
         CALL QMETRIC_COV(PAR(1), R, THETA, GD)
      CASE DEFAULT
         CALL KERR_COV(PAR(1), R, THETA, GD)
      END SELECT
   END SUBROUTINE METRIC_COV

   PURE SUBROUTINE METRIC_CONTRA(MID, PAR, R, THETA, GU, DGU_DR, DGU_DTH)
      INTEGER, INTENT(IN)   :: MID
      REAL(WP), INTENT(IN)  :: PAR(4), R, THETA
      REAL(WP), INTENT(OUT) :: GU(5), DGU_DR(5), DGU_DTH(5)
      SELECT CASE (MID)
      CASE (3)
         CALL TABLE_CONTRA(R, THETA, GU, DGU_DR, DGU_DTH)
      CASE (MID_QMETRIC)
         CALL QMETRIC_CONTRA(PAR(1), R, THETA, GU, DGU_DR, DGU_DTH)
      CASE DEFAULT
         CALL KERR_CONTRA(PAR(1), R, THETA, GU, DGU_DR, DGU_DTH)
      END SELECT
   END SUBROUTINE METRIC_CONTRA

   PURE FUNCTION INNER_BOUNDARY(MID, PAR) RESULT(RB)
      ! Radius below which a ray counts as captured: the outer event
      ! horizon for Kerr, the r = 2m singular surface for the q-metric.
      INTEGER, INTENT(IN)  :: MID
      REAL(WP), INTENT(IN) :: PAR(4)
      REAL(WP) :: RB
      SELECT CASE (MID)
      CASE (3)
         RB = PAR(1)
      CASE (MID_QMETRIC)
         RB = 2.0_WP
      CASE DEFAULT
         RB = HORIZON_RADIUS(PAR(1))
      END SELECT
   END FUNCTION INNER_BOUNDARY

END MODULE SPACETIME
